### PR TITLE
[None][feat] Support the DMD2-distilled Cosmos3-Super-Text2Image-4Step checkpoint

### PR DESCRIPTION
**Pull request by ishovkun**
## Summary by CodeRabbit

* **New Features**
  * Added support for NVIDIA Cosmos3 Super Text-to-Image, including a distilled four-step generation option.
  * Added a one-GPU 1024×1024 image-generation configuration and usage example.
  * Improved checkpoint-specific sampling, validation, warmup behavior, and image output handling.
  * Added compatibility defaults for newer Cosmos3 checkpoint configurations.

* **Documentation**
  * Documented the newly supported model, deployment configuration, usage constraints, and example command.

* **Bug Fixes**
  * Improved scheduler behavior and seeded generation consistency across supported Cosmos3 workflows.

* **Tests**
  * Added unit and end-to-end coverage for distilled generation, validation, configuration compatibility, and image output.

## Description

Adds support for [`nvidia/Cosmos3-Super-Text2Image-4Step`](https://huggingface.co/nvidia/Cosmos3-Super-Text2Image-4Step), a DMD2-distilled text-to-image Cosmos3 checkpoint. Unlike the base checkpoints, it samples with `FlowMatchEulerDiscreteScheduler` on a fixed 4-sigma stochastic (SDE) schedule declared in the checkpoint's scheduler config, with classifier-free guidance baked into the weights (one forward per step). It also ships a newer diffusers conversion of the transformer config that omits a few schema fields older conversions carried.

What changed:

- **Scheduler loading** (`sampling.py`): the pipeline instantiates the scheduler class the checkpoint declares — UniPC for base checkpoints (a missing declaration also resolves to UniPC, preserving existing behavior), FlowMatchEuler for distilled ones. An explicitly unknown declaration is a load-time error rather than a silent UniPC substitution.
- **`Cosmos3SamplingPolicy`** (`sampling.py`): an immutable value object holding the checkpoint's sampling facts (fixed sigmas, distilled detection, UniPC base config for flow-shift rebuilds). Only two recipes are valid — UniPC without fixed sigmas (base) and FlowMatchEuler with fixed sigmas plus `stochastic_sampling=true` (distilled); malformed combinations, including non-stochastic or non-SDE declarations, fail at load. Requests that conflict with a distilled checkpoint's fixed steps/guidance are rejected with a clear error, as are image-conditioned requests (correct distilled conditioning needs per-step re-anchoring, which lands with I2V-4Step support).
- **Honest generation defaults**: `default_generation_params` reports the checkpoint's true steps/guidance (4 / 1.0 for distilled). Mode-dependent fields (`height`, `width`, `num_inference_steps`, `guidance_scale`) stay `None` until `infer()` resolves the request mode (video vs. image) exactly once; explicit request values pass through unchanged.
- **Seed determinism** (`pipeline.py`): the shared denoise loop now threads `scheduler_step_kwargs` into every `scheduler.step()` call so the request-seeded `torch.Generator` drives the stochastic step's noise. Requires `diffusers>=0.39.0` — earlier versions ignore a caller-supplied generator in the stochastic branch (huggingface/diffusers#13678). Same-seed runs produce bit-identical images.
- **Transformer config schema compat** (`transformer_cosmos3.py`): newer conversions omit `position_embedding_type`, `max_position_embeddings`, and `temporal_compression_factor_sound`; these are filled with their historical values at model construction (idempotent).
- **Deployment config + docs**: `examples/visual_gen/configs/cosmos3-t2i-1gpu.yaml` warms up the deployed 1024×1024 single-frame shape (warmup follows the workflow, not the checkpoint name); README and model docs updated with the exact invocation.

Behavior for base checkpoints (Cosmos3-Nano, Cosmos3-Super) is unchanged: same UniPC scheduler, same defaults, no new step kwargs.

## Test Coverage

- `tests/unittest/_torch/visual_gen/test_cosmos3_distilled.py` (new, 69 tests): scheduler loading from real config files, recipe-matrix validation (malformed combinations fail at load), distilled request validation, flow-shift rebuild/restore semantics, fixed-sigma timestep programming, SDE seed determinism (same seed reproduces, different seeds diverge), generation defaults, `infer()` mode resolution, guidance-1.0 denoise-loop contract (single forward per step, step kwargs reach every `scheduler.step`), registry dispatch. Includes a canary pinning that diffusers retains unknown scheduler-config keys, which distilled detection depends on.
- `tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py`: config schema compat-default tests.
- `tests/integration/defs/examples/visual_gen/test_visual_gen.py::test_cosmos3_t2i_4step_example` (in `l0_b200.yml`, B200 post-merge): runs the documented example invocation against the real checkpoint and asserts an image is produced.
- Manually verified on B200: e2e image generation via the documented invocation (fixed 4-step schedule detected from the checkpoint, warmup on the 1024×1024/1-frame shape, denoise ~1.4 s), and bit-identical outputs across same-seed runs. Full cosmos3 unit suite (105 tests across the three modules) passes in a single process in both module orders.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

### CODE REVIEW

**Comment by ishovkun** [comment]
Could a maintainer please add the `api-compatible` label? No public API signature changes in this PR — the observable change is that `VisualGen.default_params` now reports `None` for mode-dependent fields (resolved per request) and the checkpoint's true steps/guidance for distilled checkpoints.

File: tests/integration/defs/examples/visual_gen/test_visual_gen.py [hunk 1]
```diff
@@ -1852,3 +1852,49 @@ def test_cosmos3_example(_visual_gen_deps, llm_root, llm_venv):
...
+            "--prompt",
+            "A ceramic teapot pouring steaming tea into a cup, morning window light",
+            "--output_type",
+            "image",
+            "--output_path",
+            output_path,
+        ],
+        env={"TRTLLM_DISABLE_COSMOS3_GUARDRAILS": "1"},
+    )
+    assert os.path.isfile(output_path), f"Example did not produce output at {output_path}"
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply tests/integration/defs/examples/visual_gen/test_visual_gen.py:2112]: Fixed in c94216137ee3c6f13507375ec85b7b05528db881 — the test now removes any pre-existing PNG before invoking the example, and asserts the fresh artifact exists and is non-empty.

File: tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py [hunk 2]
```diff
@@ -468,3 +464,32 @@ def test_load_fp8_quantization(self, quant_algo: str):
...
+            assert getattr(config, key) == value
+
+    def test_old_schema_untouched(self):
+        config = SimpleNamespace(
+            position_embedding_type="unified_3d_mrope",
+            max_position_embeddings=12345,  # deliberately non-default
+            temporal_compression_factor_sound=1,
+        )
+        apply_pretrained_config_compat_defaults(config)
+        assert config.max_position_embeddings == 12345
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py:499]: Fixed in ea72ebb2e9db9a032987c029370e990c688ff86a — all three explicitly provided fields now use non-default sentinels (`rope_3d` / `12345` / `7`), each with its own preservation assertion, so an overwrite of any one field fails observably. This test class is the only coverage of the helper, so no equivalent assertions were needed elsewhere.

File: tests/unittest/_torch/visual_gen/conftest.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    derived global (assigned by that same function), so teardown restores
+    both. Opt in per module via ``pytest.mark.usefixtures``.
+    """
+    import tensorrt_llm._torch.visual_gen.models.cosmos3.pipeline_cosmos3 as pipe_mod
+
+    patcher = pytest.MonkeyPatch()
+    patcher.setenv("TRTLLM_DISABLE_COSMOS3_GUARDRAILS", "1")
+    patcher.setattr(pipe_mod, "TRTLLM_DISABLE_COSMOS3_GUARDRAILS", True)
+    yield
+    patcher.undo()
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply tests/unittest/_torch/visual_gen/conftest.py]: Partially addressed in 860a470109ba2c466114aa1c2d9fcf236ffaca4a: the new conftest fixture is annotated (`-> Iterator[None]`) and the distilled-test helpers got precise parameter/return types (`Path`, `SimpleNamespace`).

Intentionally not applying the rest (`-> None` on every test method, fixture-parameter annotations in the integration test): the surrounding tests in those files are uniformly unannotated (including all pre-existing tests in `test_visual_gen.py`), static type checking is opt-in per submodule (CODING_GUIDELINES, Static Typing) and none of these paths are in the mypy pre-commit whitelist. Annotating only the new lines would create a style seam inside files this PR otherwise matches, with no checker to enforce it.

**Comment by ishovkun** [comment]
Responses to the two CodeRabbit items without inline threads:

**Nitpick — annotate `apply_pretrained_config_compat_defaults` with the concrete pretrained-config type** (`transformer_cosmos3.py`): addressed in e3ec19c08f1056e182d3edc7688e9a747dba2545. There is no concrete class to name — `DiffusionModelConfig.pretrained_config` is declared `Optional[Any]` and the loader builds a `SimpleNamespace` from the checkpoint JSON — so the helper is annotated with an identity `TypeVar` (`_PretrainedConfigT -> _PretrainedConfigT`), which states precisely that it mutates and returns the same configuration object without over-narrowing a deliberately structural field.

**Outside-diff — make omitted `output_type` resolve to `image`, default warmup to the T2I shape, and reject video mode for this checkpoint** (`pipeline_cosmos3.py`): not applying, by design. The checkpoint declares its *sampling recipe* (distilled FlowMatchEuler + fixed sigmas), not its modality — "distilled ⇒ image-only" does not hold (a distilled image-to-video checkpoint exists), so implementing this would require gating on the checkpoint name, which this codebase deliberately avoids. Request validation rejects what the checkpoint actually declares (conflicting steps/guidance). Warmup shape follows the deployed workflow rather than the checkpoint: `configs/cosmos3-t2i-1gpu.yaml` sets `compilation_config` to 1024×1024 / 1 frame, and the documented invocation, the README, and the B200 integration test all use it together with an explicit `--output_type image`.

**Review by xinhe-nv — APPROVED** [review]

**Review by xinhe-nv — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]
LGTM — additive DMD2-distilled Cosmos3 support; the shared pipeline.py change adds an optional scheduler_step_kwargs (default None→{}, so existing models are unchanged) and there's no public tensorrt_llm/visual_gen API change. Only cross-cutting item: the diffusers floor bump 0.37.1→0.39.0 raises the minimum for all diffusers-based VisualGen models — CI-covered, worth a heads-up.

File: tests/integration/test_lists/test-db/l0_b200.yml [hunk 4]
```diff
@@ -265,6 +265,7 @@ l0_b200:
   - examples/visual_gen/test_visual_gen.py::test_ltx2_example
   - examples/visual_gen/test_visual_gen.py::test_wan_i2v_example
   - examples/visual_gen/test_visual_gen.py::test_cosmos3_example
+  - examples/visual_gen/test_visual_gen.py::test_cosmos3_t2i_4step_example TIMEOUT (30)
```
Comment (tburt-nv) [inline tests/integration/test_lists/test-db/l0_b200.yml]: 30 minutes is a long time for a test. Can we move this to post-merge?

**Review by tburt-nv — APPROVED** [review]

File: tests/integration/test_lists/test-db/l0_b200.yml (hunk 4 above)
Comment (ishovkun) [reply tests/integration/test_lists/test-db/l0_b200.yml]: Done in f2dc31c0b8a5975b53e5b91e56ed9e07a25b6c1a — moved to the B200 post-merge section.

**Comment by ishovkun** [comment]
The L0 failure was `unittest/_torch/visual_gen/test_flux_attention.py` on `DGX_B200-PyTorch-7`. Because this PR raises the dev `diffusers` pin to 0.39.0 — which governs the version *every* VisualGen test runs against, not just the Cosmos3 ones — I checked that link directly rather than assuming the failure was unrelated:

- `test_flux_attention.py` is untouched by this PR (identical to `main`).
- Ran it on a B200 against diffusers 0.39.0: **5 passed** (`test_backend_equivalence`, `test_trtllm_backend_sanity`, `test_vanilla_backend_sanity`, `test_flux2_vanilla_backend_sanity`, plus 2 subtests).

So the 0.38 → 0.39 bump does not break Flux attention, and the failure does not appear to originate in this PR. Re-running CI with fail-fast disabled, since the multi-GPU stage was blocked by this failure last time and has no signal yet. If it reproduces, the inner assertion is in the archived `results-sub-unittests-unittest-_torch-visual_gen-test_flux_attention.py.xml` and should go to the VisualGen/Flux owners.

**Comment by ishovkun** [comment]
**Correcting my earlier comment: this failure *was* caused by this PR.** I previously said it looked unrelated — that was wrong, and here is the actual root cause, now reproduced and fixed in 14eff8312f.

The new `tests/unittest/_torch/visual_gen/conftest.py` set `TLLM_DISABLE_MPI=1` for the **entire** visual_gen unit-test directory, not just the Cosmos3 modules. `Mapping.__new__` dispatches on exactly that variable:

```python
def __new__(cls, *args, **kwargs):
    if mpi_disabled():   # TLLM_DISABLE_MPI == "1"
        return super().__new__(DeviceMeshTopology)
    else:
        return super().__new__(MpiTopology)
```

`test_flux_attention.py` constructs `Mapping(world_size=2, tp_size=2)` directly and needs `MpiTopology`, so the directory-wide set flipped its fused QK-norm + RoPE gate — producing exactly the observed `SUBFAILED(tp_size=2)` on `test_fused_qk_norm_rope_enabled_only_for_tp1`.

Reproduced on a B200, isolating the single variable:

| condition | result |
|---|---|
| `TLLM_DISABLE_MPI` unset | 5 passed |
| `TLLM_DISABLE_MPI=1` | 1 failed — `SUBFAILED(tp_size=2)` |

**Why my earlier check missed it:** I verified `test_flux_attention.py` was byte-identical to main and ran it — but from a checkout without the new conftest, so I tested the diffusers pin and never the environment change the PR actually introduces.

**Fix:** the variable is kept only where it was before this PR — module-level in the two pre-existing Cosmos3 test modules, with their teardown restored — and removed from the shared conftest. The Cosmos3 unit tests never spawn a VisualGen executor and pass without it (69 + 22 + 14 = 105 passed). With the fixed conftest collected against the built library, `test_flux_attention.py` is back to 5 passed.

**Comment by ishovkun** [comment]
Rebased onto current main (`cfebf191f3`) — the branch was 86 commits behind, and the two x86 failures (`unittest/_torch/executor`, `unittest/_torch/speculative/test_eagle3.py`) were in areas main has since fixed, e.g. #16805 (draft token accounting) and #16571 (exclude ADP/cuda-graph dummy requests from speculative-decode acceptance stats). Neither area is touched by this PR. The SBSA multi-GPU failure was infra (agent offline → 240 min Slurm walltime).

Also worth noting from the last run: `test_flux_attention.py` passed, confirming the conftest fix in the previous commit.

/bot run --disable-fail-fast

**Review by mikeiovine — APPROVED** [review]

File: examples/visual_gen/configs/cosmos3-t2i-1gpu.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# SPDX-FileCopyrightText: Copyright (c) 2022-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (chang-l) [inline examples/visual_gen/configs/cosmos3-t2i-1gpu.yaml:1]: Do we need a separate configuration file for the distilled model, or can it reuse the existing Cosmos 3 YAML configuration?

File: tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py [hunk 6]
```diff
@@ -468,3 +472,36 @@ def test_load_fp8_quantization(self, quant_algo: str):
             del pipeline
             gc.collect()
             torch.cuda.empty_cache()
+
+
+# --- CPU-only coverage: checkpoint config schema compatibility ---
+
+
+class TestConfigCompatDefaults:
```
Comment (chang-l) [inline tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py:480]: Nit: Is this test specific to the Cosmos 3 transformer? If not, could we move it to a more general unit test file, such as `test_visual_gen_params.py` or `test_utils.py`?

**Review by chang-l — APPROVED** [review]

File: examples/visual_gen/configs/cosmos3-t2i-1gpu.yaml (hunk 5 above)
Comment (ishovkun) [reply examples/visual_gen/configs/cosmos3-t2i-1gpu.yaml:1]: Good question — but it's not a *distilled-model* config, it's a **text-to-image workflow** config. Both base and distilled T2I checkpoints use it; the header names `Cosmos3-Super-Text2Image` alongside `-4Step`. Nothing in it is checkpoint-specific — steps and guidance come from the checkpoint's own scheduler config, not from YAML.

File: tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py (hunk 6 above)
Comment (ishovkun) [reply tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py:480]: It is Cosmos3-specific — `apply_pretrained_config_compat_defaults` and `PRETRAINED_CONFIG_COMPAT_DEFAULTS` are defined in `transformer_cosmos3.py` and have exactly one caller, `Cosmos3VFMTransformer.__init__` (line 726). The three fields they fill (`position_embedding_type`, `max_position_embeddings`, `temporal_compression_factor_sound`) are Cosmos3OmniTransformer checkpoint-schema keys, and the test asserts those specific values.

You're right that the mechanism is generic, though — it's a four-line "fill missing attributes from a defaults dict" loop, and FLUX solves the same problem inline at each use site (`transformer_flux.py:594-604`, ~12 `getattr(pretrained_config, key, default)` calls). A shared `apply_config_defaults(config, defaults)` helper in `visual_gen/utils.py`, with each model owning its defaults dict, would cover both. I'd rather not do that here.
